cirq/ops/raw_types.py

```diff
@@ -319,6 +321,17 @@ def _qid_shape_(self) -> Tuple[int, ...]:
         (3, 3) for a 2-qutrit ternary gate.
         """
 
+    def _commutes_with_(self, other: Any, atol: Union[int, float] = 1e-8
+                       ) -> Union[bool, NotImplementedType]:
+        if (qid_shape_protocol.qid_shape(self, None) !=
+                qid_shape_protocol.qid_shape(other, None)):
+            return NotImplemented
+        self_unitary = protocols.unitary(self, None)
```

Comment:
This fallback-to-unitary code should go into the global `cirq.commutes` method. If the objects both specify a qid_shape but it differs, they should not commute.

Reply:
This seems dangerous. For example, CNOT(a, b) and CNOT(b, a) have the same unitary. We could make sure that `GateOperation._commutes_` addresses this and that the `_commutes_` strategy precedes the unitary strategy in `cirq.commutes`, but that would be brittle.